### PR TITLE
[#1390]improvement(hive):Check whether a hive table is external when purge table

### LINKED ISSUE #1390: [Improvement] Check whether a hive table is external when purge table
### What would you like to be improved?

We shouldn't purge an external hive table in the purge table interface.

### How should we improve?

suggest add a check in `HiveCatalogOperations#purgeTable`

### PR DESCRIPTION
**Pull request by coolderli**
### What changes were proposed in this pull request?

fix  #1390  

### Why are the changes needed?

(Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  - no
  2. If you fix a bug, describe the bug.)
  - Check whether a hive table is external when purge table

Fix: #1390 

### Does this PR introduce _any_ user-facing change?
- no

### How was this patch tested?
- added unit tests

### CODE REVIEW

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java [hunk 1]
```diff
@@ -9,7 +9,9 @@
 import static com.datastrato.gravitino.catalog.hive.HiveCatalogPropertiesMeta.METASTORE_URIS;
 import static com.datastrato.gravitino.catalog.hive.HiveTable.SUPPORT_TABLE_TYPES;
 import static com.datastrato.gravitino.catalog.hive.HiveTablePropertiesMetadata.COMMENT;
+import static com.datastrato.gravitino.catalog.hive.HiveTablePropertiesMetadata.EXTERNAL;
```
Comment (mchades) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:12]: This import is not being used.

File: catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/TestHiveTable.java [hunk 2]
```diff
@@ -599,4 +602,87 @@ public void testAlterHiveTable() {
...
+                    columns,
+                    HIVE_COMMENT,
+                    properties,
+                    new Transform[0],
+                    distribution,
+                    sortOrders);
+    Assertions.assertTrue(hiveCatalog.asTableCatalog().tableExists(tableIdentifier));
+    Assertions.assertThrows(UnsupportedOperationException.class, () -> {
+      hiveCatalog.asTableCatalog().purgeTable(tableIdentifier);
+    });
```
Comment (mchades) [inline catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/TestHiveTable.java:685]: need to assert the error message too

**Review by mchades — APPROVED** [review]
Overall, the added logic is fine. BTW, Could you please add an integration test in `CatalogHiveIT` and ensure that all Github checks are passing?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/hive/CatalogHiveIT.java [hunk 3]
```diff
@@ -1105,4 +1107,54 @@ void testAlterEntityName() {
...
+            new Partitioning[] {IdentityPartitioningDTO.of(columns[2].name())});
+    // Directly get table from hive metastore to check if the table is created successfully.
+    org.apache.hadoop.hive.metastore.api.Table hiveTab =
+        hiveClientPool.run(client -> client.getTable(schemaName, tableName));
+    Assertions.assertEquals(MANAGED_TABLE.name().toLowerCase(Locale.ROOT), hiveTab.getTableType());
+    catalog
+        .asTableCatalog()
+        .purgeTable(NameIdentifier.of(metalakeName, catalogName, schemaName, tableName));
+    Boolean existed = hiveClientPool.run(client -> client.tableExists(schemaName, tableName));
+    Assertions.assertFalse(existed, "The hive table should not be exist");
```
Comment (mchades) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/hive/CatalogHiveIT.java:1130]: The HDFS file of the table should also be asserted not existing in `.Trash`. ( BTW, there is an HDFS client in this class, so you can use it directly to access HDFS :)

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/hive/CatalogHiveIT.java [hunk 4]
```diff
@@ -1105,4 +1107,54 @@ void testAlterEntityName() {
...
+        UnsupportedOperationException.class,
+        () -> {
+          catalog
+              .asTableCatalog()
+              .purgeTable(NameIdentifier.of(metalakeName, catalogName, schemaName, tableName));
+        },
+        "Can't purge a external hive table");
+
+    Boolean existed = hiveClientPool.run(client -> client.tableExists(schemaName, tableName));
+    Assertions.assertTrue(existed, "The table should be still exists");
```
Comment (mchades) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/hive/CatalogHiveIT.java:1158]: The data in the external table should also be asserted existing.

**Comment by coolderli** [comment]
@mchades 
The integration tests failed on my computer.  Any advice on how to fix it? Thanks.

Here is my operating system version. The status of the docker service is healthy. 
```
Linux version 5.4.0-169-generic (buildd@lcy02-amd64-102) (gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.2)) #187-Ubuntu SMP Thu Nov 23 14:52:28 UTC 2023

```
Exception: 

```
TrinoConnectorIT > initializationError FAILED
    java.lang.ExceptionInInitializerError
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:126)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptBeforeAllMethod(TimeoutExtension.java:68)
        at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
        at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllMethods$11(ClassBasedTestDescriptor.java:397)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllMethods(ClassBasedTestDescriptor.java:395)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:209)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:80)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.util.ArrayList.forEach(ArrayList.java:1259)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
        at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy2.stop(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
        at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

        Caused by:
        java.lang.RuntimeException: Failed to initialize ContainerSuite
            at com.datastrato.gravitino.integration.test.container.ContainerSuite.<init>(ContainerSuite.java:53)
            at com.datastrato.gravitino.integration.test.container.ContainerSuite.getInstance(ContainerSuite.java:61)
            at com.datastrato.gravitino.integration.test.trino.TrinoConnectorIT.<clinit>(TrinoConnectorIT.java:48)

            Caused by:
            org.testcontainers.containers.ContainerLaunchException: Container startup failed for image testcontainers/ryuk:0.5.1
                at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:361)
                at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:334)
                at org.testcontainers.utility.RyukResourceReaper.maybeStart(RyukResourceReaper.java:78)
                at org.testcontainers.utility.RyukResourceReaper.init(RyukResourceReaper.java:42)
                at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:231)
                at com.datastrato.gravitino.integration.test.container.ContainerSuite.<init>(ContainerSuite.java:45)
                at com.datastrato.gravitino.integration.test.container.ContainerSuite.getInstance(ContainerSuite.java:61)
                at com.datastrato.gravitino.integration.test.catalog.hive.CatalogHiveIT.<clinit>(CatalogHiveIT.java:111)
                at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.lang.reflect.Method.invoke(Method.java:498)
                at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
                at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
                at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
                at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
                at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:126)
                at org.junit.jupiter.engine.extension.TimeoutExtension.interceptBeforeAllMethod(TimeoutExtension.java:68)
                at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
                at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
                at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
                at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
                at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
                at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
                at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
                at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
                at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllMethods$11(ClassBasedTestDescriptor.java:397)
                at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
                at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllMethods(ClassBasedTestDescriptor.java:395)
                at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:209)
                at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:80)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
                at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
                at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
                at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
                at java.util.ArrayList.forEach(ArrayList.java:1259)
                at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
                at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
                at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
                at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
                at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
                at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
                at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
                at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
                at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
                at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
                at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
                at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
                at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
                at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
                at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
                at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
                at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
                at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
                at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
                at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
                at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.lang.reflect.Method.invoke(Method.java:498)
                at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
                at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
                at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
                at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
                at com.sun.proxy.$Proxy2.stop(Unknown Source)
                at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
                at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
                at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
                at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
                at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
                at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
                at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
                at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
                at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

                Caused by:
                org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
                    at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
                    at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:346)
                    ... 81 more

                    Caused by:
                    org.testcontainers.containers.ContainerLaunchException: Could not create/start container
                        at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:565)
                        at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:356)
                        at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
                        ... 82 more

                        Caused by:
                        org.testcontainers.shaded.org.awaitility.core.ConditionTimeoutException: Lambda expression in org.testcontainers.containers.GenericContainer: expected the predicate to return <true> but it returned <false> for input of <InspectContainerResponse(args=[], config=ContainerConfig(attachStderr=false, attachStdin=false, attachStdout=false, cmd=[/bin/ryuk], domainName=, entrypoint=null, env=[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin], exposedPorts=[8080/tcp], hostName=66d2610011e1, image=testcontainers/ryuk:0.5.1, labels={org.testcontainers=true, org.testcontainers.lang=java, org.testcontainers.ryuk=true, org.testcontainers.version=1.19.0}, macAddress=null, networkDisabled=true, onBuild=null, stdinOpen=false, portSpecs=null, stdInOnce=false, tty=false, user=, volumes=null, workingDir=, healthCheck=null), created=2024-01-11T02:00:00.533631794Z, driver=overlay2, execDriver=null, hostConfig=HostConfig(binds=[/var/run/docker.sock:/var/run/docker.sock:rw], blkioWeight=0, blkioWeightDevice=null, blkioDeviceReadBps=null, blkioDeviceWriteBps=null, blkioDeviceReadIOps=null, blkioDeviceWriteIOps=null, memorySwappiness=null, nanoCPUs=0, capAdd=null, capDrop=null, containerIDFile=, cpuPeriod=0, cpuRealtimePeriod=0, cpuRealtimeRuntime=0, cpuShares=0, cpuQuota=0, cpusetCpus=, cpusetMems=, devices=null, deviceCgroupRules=null, deviceRequests=null, diskQuota=null, dns=null, dnsOptions=null, dnsSearch=null, extraHosts=[], groupAdd=null, ipcMode=shareable, cgroup=, links=[], logConfig=LogConfig(type=json-file, config={}), lxcConf=null, memory=0, memorySwap=0, memoryReservation=0, kernelMemory=null, networkMode=default, oomKillDisable=false, init=null, autoRemove=true, oomScoreAdj=0, portBindings={8080/tcp=[Lcom.github.dockerjava.api.model.Ports$Binding;@2601dca0}, privileged=true, publishAllPorts=false, readonlyRootfs=false, restartPolicy=no, ulimits=null, cpuCount=0, cpuPercent=0, ioMaximumIOps=0, ioMaximumBandwidth=0, volumesFrom=[], mounts=null, pidMode=, isolation=null, securityOpts=[label=disable], storageOpt=null, cgroupParent=, volumeDriver=, shmSize=67108864, pidsLimit=null, runtime=runc, tmpFs=null, utSMode=, usernsMode=, sysctls=null, consoleSize=[0, 0], cgroupnsMode=host), hostnamePath=/var/lib/docker/containers/66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70/hostname, hostsPath=/var/lib/docker/containers/66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70/hosts, logPath=/var/lib/docker/containers/66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70/66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70-json.log, id=66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70, sizeRootFs=null, imageId=sha256:ec913eeff75a6b5ed284cf17e186213a466ec10a8f471798318ffa0813b5d828, mountLabel=, name=/testcontainers-ryuk-6e8dd20c-17fd-4534-947b-2a66ec009f3a, restartCount=0, networkSettings=NetworkSettings(bridge=, sandboxId=9027e6e8fdb3ff74c0c33544223adfa879c916ec6fdbb73e5263ab401ea88d30, hairpinMode=false, linkLocalIPv6Address=, linkLocalIPv6PrefixLen=0, ports={}, sandboxKey=/var/run/docker/netns/9027e6e8fdb3, secondaryIPAddresses=null, secondaryIPv6Addresses=null, endpointID=, gateway=, portMapping=null, globalIPv6Address=, globalIPv6PrefixLen=0, ipAddress=, ipPrefixLen=0, ipV6Gateway=, macAddress=, networks={bridge=ContainerNetwork(ipamConfig=null, links=[], aliases=null, networkID=, endpointId=, gateway=, ipAddress=, ipPrefixLen=0, ipV6Gateway=, globalIPv6Address=, globalIPv6PrefixLen=0, macAddress=)}), path=/bin/ryuk, processLabel=, resolvConfPath=/var/lib/docker/containers/66d2610011e1149770bbfc6e94406dcfb34ce80a7a1613d0d0aa7376bc087e70/resolv.conf, execIds=null, state=InspectContainerResponse.ContainerState(status=running, running=true, paused=false, restarting=false, oomKilled=false, dead=false, pid=2782236, exitCode=0, error=, startedAt=2024-01-11T02:00:07.851206492Z, finishedAt=0001-01-01T00:00:00Z, health=null), volumes=null, volumesRW=null, node=null, mounts=[InspectContainerResponse.Mount(name=null, source=/var/run/docker.sock, destination=/var/run/docker.sock, driver=null, mode=rw, rw=true)], graphDriver=GraphDriver(name=overlay2, data=GraphData(rootDir=null, deviceId=null, deviceName=null, deviceSize=null, dir=null)), platform=linux)> within 5 seconds.
                            at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
                            at org.testcontainers.shaded.org.awaitility.core.AbstractHamcrestCondition.await(AbstractHamcrestCondition.java:86)
                            at org.testcontainers.shaded.org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
                            at org.testcontainers.shaded.org.awaitility.core.ConditionFactory.until(ConditionFactory.java:691)
                            at org.testcontainers.shaded.org.awaitility.core.ConditionFactory.until(ConditionFactory.java:708)
                            at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:467)
                            ... 84 more

```

**Comment by mchades** [comment]
@xunliu 
Could you help troubleshoot the container issue above?

**Comment by mchades** [comment]
close and reopen to rerun the checks.

**Comment by mchades** [comment]
Could you please create a new issue to track the above local integration test problem? @coolderli

**Comment by mchades** [comment]
IT fails in check as:

> CatalogHiveIT > testPurgeHiveManagedTable() FAILED
>     java.io.FileNotFoundException: File hdfs://172.17.0.3:9000/user/datastrato/.Trash does not exist.
>         at org.apache.hadoop.hdfs.DistributedFileSystem.listStatusInternal(DistributedFileSystem.java:988)
>         at org.apache.hadoop.hdfs.DistributedFileSystem.access$1000(DistributedFileSystem.java:118)
>         at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1048)
>         at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1045)
>         at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
>         at org.apache.hadoop.hdfs.DistributedFileSystem.listStatus(DistributedFileSystem.java:1045)
>         at com.datastrato.gravitino.integration.test.catalog.hive.CatalogHiveIT.testPurgeHiveManagedTable(CatalogHiveIT.java:1134)
> 
> CatalogHiveIT > testPurgeHiveExternalTable() FAILED
>     org.opentest4j.AssertionFailedError: Can't purge a external hive table ==> Expected java.lang.UnsupportedOperationException to be thrown, but nothing was thrown.
>         at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:71)
>         at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:41)
>         at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3098)
>         at app//com.datastrato.gravitino.integration.test.catalog.hive.CatalogHiveIT.testPurgeHiveExternalTable(CatalogHiveIT.java:1153)

**Comment by coolderli** [comment]
> Could you please create a new issue to track the above local integration test problem? @coolderli

ok, create new issue https://github.com/datastrato/gravitino/issues/1446

**Comment by jerryshao** [comment]
@mchades can you please help @coolderli with this PR? Thanks.

**Comment by mchades** [comment]
UT fails as:
> > Task :server:test
> 
> TestUtils > testUnsupportedOperation() FAILED
>     org.opentest4j.AssertionFailedError: expected: <Invalid argument> but was: <Unsupported operation>
>         at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
>         at app//org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
>         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
>         at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
>         at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1141)
>         at app//com.datastrato.gravitino.server.web.TestUtils.testUnsupportedOperation(TestUtils.java:130)

**Comment by coolderli** [comment]
@mchades Why does the CI pipeline ask for approval every time? Is that because this is my first merge request?

**Comment by qqqttt123** [comment]
> @mchades Why does the CI pipeline ask for approval every time? Is that because this is my first merge request?

Yes, because you're the first-time contributor.

**Comment by coolderli** [comment]
@jerryshao @mchades @qqqttt123 The ci pipeline is passed, could you please review it again? thanks

**Review by mchades — APPROVED** [review]
LGTM

**Pull request merged**